keyboards/keychron/v4/iso/keymaps/mischback/keymap.c

```diff
@@ -0,0 +1,338 @@
+static void personal_base_layer_lighting(void);
+static void personal_layer_indicator(uint8_t red, uint8_t green, uint8_t blue);
+#endif // RGB_MATRIX_ENABLE
+bool process_record_user(uint16_t keycode, keyrecord_t *record);
+bool rgb_matrix_indicators_user(void);
+
+
+#ifdef RGB_MATRIX_ENABLE
+/* Control the lighting of the base layer.
+ *
+ * The lighting modes of the base layer may be cycled using the ``RGB_MOD`` key
+ * and include several presets.
+ *
+ * TODO: Define/implement application-specific layers:
+ *         - GIMP
+ *         - Inkscape
+ *         - KiCAD
+ *         - Blender
+ */
+static void personal_base_layer_lighting(void) {
+
+    // Start with all lights off and then activate only the desired keys/LEDs
+    // Note: BASE_MODE_DARK is effectively this, so it doesn't need to be
+    //       handled in a dedicated case.
+    rgb_matrix_set_color_all(RGB_OFF);
+
+    switch(base_layer_mode) {
+        case BASE_MODE_DEFAULT:
+            // Just the coherent indicators (ESC/RCTL)
+            personal_layer_indicator(COLOR_BASE);
+            break;
+        case BASE_MODE_BRIGHT:
+            // Shine bright like a diamond, but give the homing keys another
+            // color.
+            rgb_matrix_set_color_all(COLOR_BASE);
+            rgb_matrix_set_color(LED_KEY_F, RGB_RED);
+            rgb_matrix_set_color(LED_KEY_J, RGB_RED);
+            break;
+        default:
+            break;
+    }
+}
+
+
+/* Activate pre-defined LEDs on every layer. */
+static void personal_layer_indicator(uint8_t red, uint8_t green, uint8_t blue) {
+    rgb_matrix_set_color(LED_KEY_ESC, red, green, blue);
+    rgb_matrix_set_color(LED_KEY_RCTL, red, green, blue);
+}
+#endif // RGB_MATRIX_ENABLE
+
+
+/* Custom keyboard initialization code. */
+void keyboard_post_init_user(void) {
+
+#ifdef RGB_MATRIX_ENABLE
+    /* Set a common baseline for the lighting.
+     *
+     * Ref: https://docs.qmk.fm/#/feature_rgb_matrix?id=indicators-without-rgb-matrix-effect
+     */
+    rgb_matrix_mode_noeeprom(RGB_MATRIX_SOLID_COLOR);
+    rgb_matrix_sethsv_noeeprom(HSV_OFF);
+#endif // RGB_MATRIX_ENABLE
+}
+
+
+/* Custom processing of keycodes.
+ *
+ * Currently this function is only used to support the custom implementation
+ * of RGB lighting.
+ *
+ * ``RGB_TOG`` is supposed to toggle the LEDs on/off, but this is not working
+ * as expected. Instead, the (keymap-specific) flag ``lighting_on`` is used to
+ * keep track of the lighting status, which is then picked up in
+ * ``rgb_matrix_indicators_user()`` below.
+ * This does work well enough for now.
+ *
+ * My rough guess would be, that this is not working because I messed something
+ * up in my ``config.h``.
+ */
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    switch(keycode) {
+#ifdef RGB_MATRIX_ENABLE
+        case RGB_TOG:
+            // RGB_TOG is not working as expected. Nonetheless, pass the
+            // keycode down the line for further processing.
```

Comment:
Thanks for the feedback!

If I remove the whole ``process_record_user()`` function (as it only deals with the RGB lightning), ``RGB_TOG`` has no effect (or I get the [docs](https://docs.qmk.fm/#/keycodes?id=rgb-lighting) wrong).

The desired outcome would be: when LEDs are active and the button is pressed, all lights are turned *off*. Pressing again turns them *on*. But it seenms, that the button/keycode has no effect.

This might be related to ``rgb_matrix_indicators_user()``?

Reply:
There is some special handling in the v4 code for RGB_TOG.  It should be setting the flags and allowing it to render the indicators even when it's off.
https://github.com/drashna/qmk_firmware/blob/422d502903c43c19007c95e6206d34f22da63092/keyboards/keychron/v4/v4.c#L44-L62

If you want it to not, then call `rgb_matrix_toggle()` instead, and return false.

Reply:
> If you want it to not, then call `rgb_matrix_toggle()` instead, and return false.

Thank you for that hint! It works now as expected!